src/graph.rs

```diff
@@ -532,12 +532,30 @@ impl Serialize for Value {
             Value::Boolean(value) => serializer.serialize_bool(*value),
             Value::Integer(value) => serializer.serialize_u32(*value),
             Value::String(value) => serializer.serialize_str(value),
-            // FIXME: there's no way to distinguish sets and lists, so we can't roundtrip accurately
-            Value::List(list) => serializer.collect_seq(list),
-            Value::Set(set) => serializer.collect_seq(set),
-            // FIXME: we don't distinguish between syntax tree node IDs, graph node IDs, and integers
-            Value::SyntaxNode(node) => serializer.serialize_u32(node.index),
-            Value::GraphNode(node) => serializer.serialize_u32(node.0),
+            Value::List(list) => {
+                let mut map = serializer.serialize_map(None)?;
+                map.serialize_entry("type", "list")?;
+                map.serialize_entry("value", list)?;
+                map.end()
+            }
+            Value::Set(set) => {
+                let mut map = serializer.serialize_map(None)?;
+                map.serialize_entry("type", "set")?;
+                map.serialize_entry("value", set)?;
+                map.end()
+            }
+            Value::SyntaxNode(node) => {
+                let mut map = serializer.serialize_map(None)?;
+                map.serialize_entry("type", "syntaxNode")?;
+                map.serialize_entry("value", &node.index)?;
+                map.end()
+            }
+            Value::GraphNode(node) => {
+                let mut map = serializer.serialize_map(None)?;
+                map.serialize_entry("type", "graphNode")?;
+                map.serialize_entry("value", &node.0)?;
```

Comment:
In a similar vain, I'd suggest `id` here if the graph nodes have an `id` field, or `index` if there live in an array and their position is used to identify them.